components/webext-storage/src/db.rs

```diff
@@ -0,0 +1,270 @@
+                } else {
+                    Err(e)
+                }
+            }
+        }
+    }
+
+    /// Open a read-only connection to the database.
+    pub fn open_read_connection(&self) -> Result<Connection> {
+        let conn = Connection::open_with_flags(self.db_path.clone(), get_read_flags())?;
+        init_sql_connection(&conn, false)?;
+        Ok(conn)
+    }
+
+    /// Close a connection to the database. If the connection is the write
+    /// connection, you will have closed the only connection able to write, so
+    /// don't do that unless you are sure.
+    pub fn close_connection(&self, _conn: Connection) -> Result<()> {
+        Ok(())
+    }
+}
+
+fn init_sql_connection(conn: &Connection, is_writable: bool) -> Result<()> {
+    let initial_pragmas = "
+        -- Cargo-culted from places.
+        PRAGMA page_size = 32768;
+
+        -- Disable calling mlock/munlock for every malloc/free.
+        -- In practice this results in a massive speedup, especially
+        -- for insert-heavy workloads.
+        -- XXX - is this relevant?
+        PRAGMA cipher_memory_security = false;
+
+        -- `temp_store = 2` - also cargo-culted - See both places and
+        -- https://github.com/mozilla/mentat/issues/505.
+        PRAGMA temp_store = 2;
+
+        -- Moar cargo-cult.
+        PRAGMA cache_size = -6144;
+
+        -- We probably will end up needing foreign-key support
+        -- PRAGMA foreign_keys = ON;
+
+        -- we unconditionally want write-ahead-logging mode
+        PRAGMA journal_mode=WAL;
+
+        -- How often to autocheckpoint (in units of pages).
+        -- 2048000 (our max desired WAL size) / 32760 (page size).
+        PRAGMA wal_autocheckpoint=62
+    ";
+
+    conn.execute_batch(initial_pragmas)?;
+    define_functions(&conn)?;
+    conn.set_prepared_statement_cache_capacity(128);
+    if is_writable {
+        let tx = conn.unchecked_transaction()?;
+        schema::init(&conn)?;
+        tx.commit()?;
+    };
+    Ok(())
+}
+
+fn define_functions(_c: &Connection) -> Result<()> {
+    Ok(())
+}
+
+// These should be somewhere else...
+pub fn put_meta(db: &Connection, key: &str, value: &dyn ToSql) -> Result<()> {
+    db.conn().execute_named_cached(
+        "REPLACE INTO moz_meta (key, value) VALUES (:key, :value)",
+        &[(":key", &key), (":value", value)],
+    )?;
+    Ok(())
+}
+
+pub fn get_meta<T: FromSql>(db: &Connection, key: &str) -> Result<Option<T>> {
+    let res = db.conn().try_query_one(
+        "SELECT value FROM moz_meta WHERE key = :key",
+        &[(":key", &key)],
+        true,
+    )?;
+    Ok(res)
+}
+
+pub fn delete_meta(db: &Connection, key: &str) -> Result<()> {
+    db.conn()
+        .execute_named_cached("DELETE FROM moz_meta WHERE key = :key", &[(":key", &key)])?;
+    Ok(())
+}
+
+// Utilities for working with paths.
+// (From places_utils - ideally these would be shared, but the use of
+// ErrorKind values makes that non-trivial.
```

Comment:
I think we could share these if we made them generic over anything that can convert from a parse error, something like `fn ensure_url_path<E: From<url::ParseError>>(p: impl AsRef<Path>) -> Result<Url, E>`. But good fodder for a follow-up (good second bug? 😄)!